File: pixsim7/backend/main/shared/namespaced_id.py
"""
Canonical namespaced identifier utilities.

Convention: ``namespace:name`` where namespace is the part before the
**first** colon and name is everything after it.  This means names
can contain colons (e.g., ``set:thief:01`` → namespace="set", name="thief:01").

This module is the single source of truth for parsing/building
colon-delimited identifiers across the backend codebase.
"""


def parse_namespaced_id(slug: str) -> tuple[str, str]:
    """
    Parse a ``namespace:name`` string by splitting on the first colon.

    Raises ``ValueError`` if the string has no colon or either part is empty.

    Examples::

        parse_namespaced_id("character:alice")   # ("character", "alice")
        parse_namespaced_id("set:thief:01")      # ("set", "thief:01")
        parse_namespaced_id("scene:game:123")    # ("scene", "game:123")
        parse_namespaced_id("bare-name")         # ValueError
        parse_namespaced_id(":oops")             # ValueError
    """
    colon_index = slug.find(":")
    if colon_index <= 0:
        raise ValueError(
            f"Invalid namespaced ID: {slug!r}. "
            "Expected 'namespace:name' with a non-empty namespace."
        )

    namespace = slug[:colon_index]
    name = slug[colon_index + 1:]
    if not name:
        raise ValueError(
            f"Invalid namespaced ID: {slug!r}. "
            "Name part (after colon) cannot be empty."
        )

    return namespace, name


def make_namespaced_id(namespace: str, name: str) -> str:
    """
    Build a ``namespace:name`` string from parts.

    Examples::

        make_namespaced_id("character", "alice")  # "character:alice"
        make_namespaced_id("set", "thief:01")     # "set:thief:01"
    """
    return f"{namespace}:{name}"


def get_namespace(slug: str) -> str | None:
    """
    Extract just the namespace from a ``namespace:name`` string.

    Returns ``None`` if the string has no colon or the namespace is empty.
    """
    colon_index = slug.find(":")
    if colon_index <= 0:
        return None
    return slug[:colon_index]

File: packages/shared/helpers/core/src/namespacedId.ts
/**
 * Canonical namespaced identifier utilities.
 *
 * Convention: `namespace:name` where namespace is the part before the
 * **first** colon and name is everything after it. This means names
 * can contain colons (e.g., `set:thief:01` → namespace="set", name="thief:01").
 *
 * This module is the single source of truth for parsing/building
 * colon-delimited identifiers across the frontend codebase.
 */

export interface NamespacedId {
  namespace: string;
  name: string;
}

/**
 * Parse a `namespace:name` string by splitting on the first colon.
 *
 * Returns `null` if the string has no colon or either part is empty.
 *
 * @example
 * parseNamespacedId("character:alice")   // { namespace: "character", name: "alice" }
 * parseNamespacedId("set:thief:01")      // { namespace: "set", name: "thief:01" }
 * parseNamespacedId("scene:game:123")    // { namespace: "scene", name: "game:123" }
 * parseNamespacedId("bare-name")         // null
 * parseNamespacedId(":oops")             // null
 */
export function parseNamespacedId(id: string): NamespacedId | null {
  const colonIndex = id.indexOf(':');
  if (colonIndex <= 0) return null; // no colon, or colon at position 0 (empty namespace)

  const namespace = id.slice(0, colonIndex);
  const name = id.slice(colonIndex + 1);
  if (!name) return null; // empty name

  return { namespace, name };
}

/**
 * Build a `namespace:name` string from parts.
 *
 * @example
 * makeNamespacedId("character", "alice")     // "character:alice"
 * makeNamespacedId("set", "thief:01")        // "set:thief:01"
 */
export function makeNamespacedId(namespace: string, name: string): string {
  return `${namespace}:${name}`;
}

/**
 * Extract just the namespace from a `namespace:name` string.
 *
 * Returns `null` if the string has no colon or the namespace part is empty.
 *
 * @example
 * getNamespace("character:alice")  // "character"
 * getNamespace("set:thief:01")     // "set"
 * getNamespace("bare-name")        // null
 */
export function getNamespace(id: string): string | null {
  const colonIndex = id.indexOf(':');
  if (colonIndex <= 0) return null;
  return id.slice(0, colonIndex);
}

/**
 * Extract just the name from a `namespace:name` string.
 *
 * Returns `null` if the string has no colon or the name part is empty.
 *
 * @example
 * getName("character:alice")  // "alice"
 * getName("set:thief:01")     // "thief:01"
 * getName("bare-name")        // null
 */
export function getName(id: string): string | null {
  const colonIndex = id.indexOf(':');
  if (colonIndex <= 0) return null;
  const name = id.slice(colonIndex + 1);
  return name || null;
}
